### PR TITLE
ARROW-9453: [Rust] Wasm32 compilation support

### PR DESCRIPTION
**Pull request by ghost**
Adding compilation flag to allow rust library to compile against Wasm32 target

### CODE REVIEW

File: rust/arrow/src/util/bit_util.rs [hunk 1]
```diff
@@ -366,13 +371,13 @@ mod tests {
         assert_eq!(ceil(8, 8), 1);
         assert_eq!(ceil(9, 8), 2);
         assert_eq!(ceil(9, 9), 1);
-        assert_eq!(ceil(10000000000, 10), 1000000000);
-        assert_eq!(ceil(10, 10000000000), 1);
-        assert_eq!(ceil(10000000000, 1000000000), 10);
+        assert_eq!(ceil(10000000, 10), 1000000);
```
Comment (paddyhoran) [inline rust/arrow/src/util/bit_util.rs:374]: Why the changes here?

**Comment by paddyhoran** [comment]
@rj-atw I would like to see us get wasm support but I think we need to add CI for this.  Without CI all those `cfg` attributes are destined to go stale...

**Comment by paddyhoran** [comment]
Also, I assigned the JIRA to you and changed your permissions so you can self-assign in future

File: rust/arrow/src/util/bit_util.rs (hunk 1 above)
Comment (rj-atw) [reply rust/arrow/src/util/bit_util.rs:374]: I will double check these changes. I made them in response to a test that was failing; however, I found it suspect at the time as well.

**Comment by rj-atw** [comment]
> 
> 
> @rj-atw I would like to see us get wasm support but I think we need to add CI for this. Without CI all those `cfg` attributes are destined to go stale...

@paddyhoran I love to add the need CI. Do we have any documentation around this projects CI (especially how to setup env locally)?

File: rust/Cargo.toml [hunk 2]
```diff
@@ -24,3 +24,7 @@ members = [
         "integration-testing",
 	"benchmarks",
 ]
+default-members= [
```
Comment (sunchao) [inline rust/Cargo.toml:27]: Why add this? is this related to the change?

**Review by sunchao — COMMENTED** [review]
It will be good to add more details in the PR description, e.g., what have been disabled for wasm32 and what are the consequences etc.

File: rust/arrow/src/lib.rs [hunk 3]
```diff
@@ -31,17 +31,19 @@ pub mod array;
 pub mod bitmap;
 pub mod buffer;
 pub mod compute;
+#[cfg(not(target_arch="wasm32"))]
 pub mod csv;
 pub mod datatypes;
 pub mod error;
-#[cfg(feature = "flight")]
+#[cfg(all(feature = "flight",not(target_arch="wasm32")))]
```
Comment (vertexclique) [inline rust/arrow/src/lib.rs:38]: Inconsistent spacing between macro features and selections. Please reorganize them consistently.

Comment (vertexclique) [reply rust/arrow/src/lib.rs:38]: Also please run cargo fmt and cargo clippy

File: rust/datafusion/Cargo.toml [hunk 4]
```diff
@@ -46,18 +46,20 @@ cli = ["rustyline"]
...
 fnv = "1.0"
 arrow = { path = "../arrow", version = "1.0.0-SNAPSHOT" }
-parquet = { path = "../parquet", version = "1.0.0-SNAPSHOT" }
 sqlparser = "0.2.6"
-clap = "2.33"
-prettytable-rs = "0.8.0"
+paste = "0.1"
+
+[target.'cfg(not(target_arch="wasm32"))'.dependencies]
+parquet = { path = "../parquet", version = "1.0.0-SNAPSHOT" }
```
Comment (vertexclique) [inline rust/datafusion/Cargo.toml:53]: Don't not gate parquet if not wasm. The parquet should be also wasm compilable. Moreover, actual dependencies shouldn't be gated under the negated set, which makes it harder to maintain.

File: rust/datafusion/src/bin/repl.rs [hunk 5]
```diff
@@ -17,15 +17,29 @@
...
 use std::env;
+#[cfg(not(target_arch="wasm32"))]
 use std::path::Path;
+#[cfg(not(target_arch="wasm32"))]
 use std::time::Instant;
 
+
+#[cfg(target_arch="wasm32")]
+pub fn main() {
+}
```
Comment (vertexclique) [inline rust/datafusion/src/bin/repl.rs:40]: What is the point of not compiling the ported code? I don't see the benefit of porting to wasm by doing this.

**Review by vertexclique — CHANGES_REQUESTED** [review]
Made a couple of comments. Nice to see this opened!

File: rust/Cargo.toml (hunk 2 above)
Comment (paddyhoran) [reply rust/Cargo.toml:27]: Is this just so that wasm builds by default at the top level?  If so, I'd rather not do this either.  i.e. it's enough to just say that we can build the sub-crates on wasm instead of changing this global default.

**Comment by paddyhoran** [comment]
> @paddyhoran I love to add the need CI. Do we have any documentation around this projects CI (especially how to setup env locally)?

If you search for "rust" under the top level "ci" folder you will find most of the CI scripts.  We can then ask @kszucs for a quick review of that part of the update.

**Comment by kszucs** [comment]
@rj-atw you can start here https://github.com/apache/arrow/blob/master/docs/source/developers/docker.rst

I assume you'll need to adjust the [rust_build.sh](https://github.com/apache/arrow/blob/master/ci/scripts/rust_build.sh) script to accept either a target arch environment variable (preferable) or a CLI argument than add a new docker-compose entry (like debian-rust-wasm32) similar to [debian-rust](https://github.com/apache/arrow/blob/master/docker-compose.yml#L928) where you pass the wasm variable.

Once that works with `archery docker run debian-rust-wasm32` you can add a matrix entry to execute debian-rust-wasm32 in the [rust github actions configuration](https://github.com/apache/arrow/blob/master/.github/workflows/rust.yml#L52).

File: rust/arrow/src/compute/kernels/arithmetic.rs [hunk 6]
```diff
@@ -99,8 +99,8 @@ where
 }
 
 /// SIMD vectorized version of `math_op` above.
-#[cfg(all(any(target_arch = "x86", target_arch = "x86_64"), feature = "simd"))]
-fn simd_math_op<T, F>(
+#[cfg(all(any(target_arch = "x86", target_arch = "x86_64", target_arch="wasm32"), feature = "simd"))]
+pub fn simd_math_op<T, F>(
```
[the remark this answers is not in the document]
Comment (rj-atw) [reply rust/arrow/src/compute/kernels/arithmetic.rs:103]: Thanks. I made this modification to explicitly test that simd op call was not failing in WASM. Forgot to revert while pushing

File: rust/Cargo.toml (hunk 2 above)
Comment (rj-atw) [reply rust/Cargo.toml:27]: Yes. This was add as I did not try and port the Flight and Parquet submodules as I believe their core functionality uses file and sockets which don't have stable support in WASM yet. 

I'll try out the individual submodule build, add details to readme and revert the Cargo.toml change.

File: rust/datafusion/Cargo.toml (hunk 4 above)
Comment (rj-atw) [reply rust/datafusion/Cargo.toml:53]: Will look more into parquet library. I skipped many as it is primarily used to read from the file system...which I believe does not have stable support yet in WASM.

If I change to use a positive gate should I restrict to just x86 and x86_64; or are there other supported targets?

File: rust/datafusion/src/bin/repl.rs (hunk 5 above)
Comment (rj-atw) [reply rust/datafusion/src/bin/repl.rs:40]: WASM is a library only. I believe this code is only useful when creating an executable. 

I know this is a bad hack, but I am actually not really sure why this file was still pickup by rustc as I should be excluding in the package dependencies. 

I'll take another look at why this is still being compile. However, I am open to suggestion/discussion on alternatives.

**Comment by emkornfield** [comment]
@rj-atw where you able to figure out CI?  It looks like this also needs to be rebased.

File: rust/arrow/src/util/bit_util.rs (hunk 1 above)
Comment (nevi-me) [reply rust/arrow/src/util/bit_util.rs:374]: The numbers are 64-bit, which won't be supported in wasm32

Comment (alippai) [reply rust/arrow/src/util/bit_util.rs:374]: I think wasm32 supports i64 and f64 https://webassembly.github.io/spec/core/syntax/types.html

**Comment by nevi-me** [comment]
Hi @rj-atw, this PR has fallen behind quite a lot. Are you still interested in or able to work on this feature? Otherwise we could close the PR in the interim.

**Comment by jorgecarleitao** [comment]
@rj-atw I am adding wasm32 compilation to the CI, so that this can progress. Is this something that you are still interested on, or should we close this PR?

**Comment by rj-atw** [comment]
I am still intersted.

Started new job, so diverted my attention. But hoping to close this out
during holiday break.

**Comment by alamb** [comment]
@rj-atw  -- FYI @jorgecarleitao  has the appropriate CI support here: https://github.com/apache/arrow/pull/8864

**Comment by alamb** [comment]
@rj-atw  --  Given the imminent Arrow 3.0 release, I am trying to clean up older Rust PRs and see if the authors have plans to move them forward. 

Do you plan on working on this PR in the near future? If not, should we close this PR until there is time to make progress? Thanks again for your contributions so far.

**Comment by nevi-me** [comment]
I'm closing this for now @alamb @rj-atw, we can pick it up later.
